코드 구조 리팩토링

namsic opened the pull request:
### 변경사항
- 에러 응답과 정상 응답 모두 동일한 format(ResponseDTO) 사용하도록 변경

- 200, 400 외의 http status code 사용

- 에러 응답에 code 추가

- Layer 별 역할 구분
  - Controller에서는 요청 확인 & 응답 생성 위주로 동작
  - 주요 business logic은 Service layer 호출하여 처리

- Member Entity 변경
  - json property 제거: 굳이 snake case 써야하는 이유를 잘 모르겠음
  - Validation annotation 최적화

- JpaRepository 사용

- 위 변경으로 사용되지 않게 된 파일 제거

- 테스트코드 변경

### 추후 수정 필요
- 주석 추가
- 하드코딩 제거
- API 문서 변경

namsic committed:
refactor: Update Response and Member API

namsic committed:
test: Update test for member api

namsic committed:
minor change

B2SIC commented:
## 리뷰

### Repository 관련 내용
먼저 Spring Data JPA를 사용하여 Repository를 구성하였는데, 이전 PR에서 Spring Data JPA는 JPA의 이해를 기반으로 많은 부분이 함축된 채로 기능을 사용 해야 하기 때문에 JPA로 우선 진행을 하겠다고 언급을 했었습니다.

이 부분을 고려하신 것인지, 그럼에도 Spring Data JPA 사용을 권하시는지 궁금합니다.

관련해서 회원 정보를 검색하는 기능의 Query문이 직접 작성이 되었는데
```
@Query("SELECT m FROM Member m WHERE 1 = 1"
         + " AND (:userId is null OR m.userId = :userId)"
         + " AND (:phoneNumber is null OR m.phoneNumber = :phoneNumber)")
```

현재 모든 회원 정보를 리턴하고 있다는 문제가 있습니다.
또한 필요한 정보만 Return 하는 코드가 구조가 변경되면서 삭제된 것으로 보여 id key 값과 password 데이터도 함께 노출되고 있습니다.

### ResponseDTO
코드 및 구조가 명확하고 간결해진 것 같아서 좋은 것 같습니다.
정상 데이터 반환과 에러 응답을 ResponseDTO로 함께 사용하는 방법은 초기에 사용했던 방식인데 처리 해야 할 Exception이 많아지면서 방식을 변경했었습니다. 하나로 깔끔하게 통합된 방식을 보니 훨씬 나아진 것 같습니다.

다만 정상 응답의 경우 error 필드를, 에러 상황에서는 data 필드를 숨기는 것이 더 좋지 않을까요? 관련 코드가 삭제되었길래 질문 남깁니다.

### Exception Handler
변경 사항 관련 테스트를 하면서 발생했던 문제는 입력 데이터 관련 오류가 있을 때 처리되던 전역 Exception 및 커스텀 Exception이 모두 사라진 상태여서 입력 데이터에 문제가 생길 때 발생하던 HttpMessageNotReadableException을 처리 못하고 있습니다.

또한 현재 Exception이 BindingResult에 에러 데이터가 없으면 null을 Return 하도록 구성이 되어있는데, BindingResult에 데이터가 없어도 에러가 발생할 수 있는 상황이 존재하는데 이런 상황 역시 BusinessException으로 핸들링 하도록 의도한 것인가요?

Exception 구조 역시 뭔가 사용하기 편리해진 것으로 보입니다.
직접 사용해보면서 필요에 따라 추가 및 수정을 해도 괜찮을 것으로 보입니다.

### 기타
전반적으로 핵심 기능을 담당하는 코드 내용부터 구조까지 전부 바뀌어서 리팩토링이라기보다는 새로 짠 코드에 더 가까운 것 같습니다.
그만큼 변경 사항이 많아서 꾸준하게 정교한 테스트가 필요할 것으로 보입니다.
추가로 FIXME 라고 되어있는 주석 중에 수정이 끝난 주석은 제거 부탁 드립니다. 제가 작업 할 대상인지 작업이 완료된 대상인지 구분이 잘 안갑니다. 또한 새롭게 작성된 클래스에 대한 주석은 작성자가 작성해주는 것이 의도 파악 측면에서 훨씬 효율적일 것으로 생각됩니다.

다음 작업으로 진행하기 이전에 동일 작업 반복을 피하기 위해서 리팩토링 코드에 대한 리뷰 및 반영 여부 결정이 필요해보입니다.

namsic commented:
@B2SIC 
### JPA 관련

> 이전 PR에서 Spring Data JPA는 JPA의 이해를 기반으로 많은 부분이 함축된 채로 기능을 사용 해야 하기 때문에 JPA로 우선 진행을 하겠다고 언급을 했었습니다.

본 리팩토링은 반드시 적용되어야 할 코드가 아닌 의견 제시의 성격이 강하므로
개인적으로 편리해 보여 적용한 것이며 반드시 권하는 것은 아닙니다.
`JpaRepository`를 적용한 나름의 이유는 아래와 같습니다.
- 현재 질의문은 아주 단순한 형태를 띄고 있으며 이후로도 복잡한 쿼리는 없을 것 같은데,
JPA에 대한 이해가 부족해서 문제가 생기는 상황이 무엇인지 예상 불가
- 이전 리뷰에서 `findByValue()`를 사용해 추상화하여 코드 중복을 한 단계 줄였는데,
`JpaRepository`를 사용하면 선언만으로 사용 가능하여 더 간결한 코드로 변경 가능
- 이미 잘 만들어진 것을 두고 처음부터 구현할 이유가 없다고 판단

본 프로젝트는 이론 학습보다는 포트폴리오 구성에 조금 더 비중을 두고 있으므로
실제로 현업에서 여러 방식 중 무엇을 더 많이 사용하는지 알 수 있다면
그 방법을 적용하여 익숙해지는 것이 좋겠습니다.

> 회원 정보를 검색하는 기능의 Query문이 직접 작성이 되었는데 현재 모든 회원 정보를 리턴하고 있다는 문제가 있습니다.

- GET `/member`
모든 member 조회
- GET `/member?fieldA=value1`
`fieldA`가 `value1`인 member 조회
- GET `/member?fieldA=value1&fieldB=value2`
`fieldA`가 `value1`이고 `fieldB`가 `value2`인 member 조회

위 동작이 RESTful한 것 같아 사용했는데, 의도와 다르게 동작하고 있나요?
아니면 설계 관점에서 문제가 있나요?

### Exception 관련

> 입력 데이터에 문제가 생길 때 발생하던 HttpMessageNotReadableException을 처리 못하고 있습니다.

`MethodArgumentNotValidException`과 어떻게 다른가요? 어떤 상황에서 발생하는지 잘 모르겠어서 추가하지 않았는데, 일부러 제외한 것은 아니며 필요한 경우 추가해야 합니다.

> BindingResult에 데이터가 없어도 에러가 발생할 수 있는 상황이 존재하는데 이런 상황 역시 BusinessException으로 핸들링 하도록 의도한 것인가요?

어떤 상황인지 예시를 들어줄 수 있나요? 이와 관련해서 별도로 의도한 것은 없습니다.

### 기타
> 다만 정상 응답의 경우 error 필드를, 에러 상황에서는 data 필드를 숨기는 것이 더 좋지 않을까요? 관련 코드가 삭제되었길래 질문 남깁니다.

통일된 구조를 가지도록 구현한 것은 아래의 이유에 따라 의도한 것입니다.
- 모든 응답이 일관된 구조를 가지게 하기 위함
- 성공/실패 여부는 status code로 판단하므로 특정 field의 존재 여부로 구분할 필요 X
- 필요하다면 추가 정보를 담아 전달할 수 있을 것으로 보임
  - 성공 시 error에 warning 정보를 전달
    - error를 전혀 신경쓰지 않고 data만 사용해도 무방
    - warning 정보를 확인하여 그에 따른 동작
  - 실패 시 data를 전달
  (만약 요청 data 중 일부만 조회에 성공한 경우, 일부 data만 전달)
    - data는 무시하고 error만 확인하여 조치 후 올바른 새 요청 전송
    - 일부 data라도 클라이언트에게 의미가 있는 경우, 이 값을 사용 가능

마찬가지로 하나의 의견일 뿐이니 검토 바랍니다.

> - FIXME 라고 되어있는 주석 중에 수정이 끝난 주석은 제거 부탁 드립니다.
> - 필요한 정보만 Return 하는 코드가 구조가 변경되면서 삭제된 것으로 보여 id key 값과 password 데이터도 함께 노출되고 있습니다.

FIXME 주석은 모두 수정이 필요한 내용입니다.
다만 현재는 id와 pw가 노출되는 것에(적절하지는 않지만) 큰 문제가 없고
(pw는 raw data가 아닌 해시값 사용 예정이고, id는 노출되면 어떤 문제가 있는지?)
나머지 항목 또한 프로그램 주요 흐름에 영향을 주는 내용은 아니라고 판단하여
FIXME를 사용하여 메모해두었습니다.

B2SIC reviewed (commented):
## JPA 관련
> 본 프로젝트는 이론 학습보다는 포트폴리오 구성에 조금 더 비중을 두고 있으므로
실제로 현업에서 여러 방식 중 무엇을 더 많이 사용하는지 알 수 있다면
그 방법을 적용하여 익숙해지는 것이 좋겠습니다.

이러한 이유라면 Spring Data JPA를 사용하는 것도 고려해보겠습니다.
다만 현재의 질의문이 대상이 아닌 앞으로 복잡해질 상황들을 우려해서 언급한 내용이었습니다.

> 위 동작이 RESTful한 것 같아 사용했는데, 의도와 다르게 동작하고 있나요?
아니면 설계 관점에서 문제가 있나요?

의도와 다르게 동작하고 있습니다.
API 테스트를 해보면 쿼리에 user_id 또는 phone_number 존재 여부와 상관없이 항상 모든 데이터를 반환합니다.
그리고 전체 회원 조회 API는 제공할 필요가 없다고 생각했는데 필요한 상황이 있을까요?

## Exception
`MethodArgumentNotValidException` 은 Spring의 `@Valid` 어노테이션을 통해 사전에 정의된 제약 조건을 지키지 않았을 경우 발생할 수 있습니다.
`HttpMessageNotReadableException` 은 말 그대로 입력 데이터를 읽을 수 없어 요청을 처리할 수 없는 상황에서 발생할 수 있습니다. 그 예로 제시했던 것이 JSON Type이 아닌 요청이었습니다.

## 기타
의도와 이유 모두 동의합니다.
번거롭겠지만 JPA 관련 내용에 대해서 다시 검토 부탁드립니다.
반영할 수 있는 부분들을 빠르게 반영하고 다음 작업으로 넘어갈 수 있도록 하겠습니다.

namsic commented:
@B2SIC
### 메인 개발자
`JPA의 이해를 기반으로 많은 부분이 함축된 채로 기능을 사용`하면 어떤 문제가 생기는지 알지 못하므로 JpaRepository 적용 시의 코드 구조를 보여주고 이에 대한 의견을 묻기 위해 PR에 포함시켰지만, 최종 결정은 main developer가 직접 해 주었으면 합니다.

리뷰 및 PR 작성 과정에서 Spring에 대해 별도로 공부하지 않고 작성한 코드를 보면서 필요한 부분만 검색해 알아보고, 나름대로 이해한 내용을 가지고 더 낫다고 생각하는 형태를 제안합니다. 따라서 작성한 코드가 항상 맞는 코드라고 보장할 수 없습니다.
예를 들어, 아래 상황은 특정 의도를 가지고 변경한 것이 아니라 단순히 배경 지식이 부족하여 발생한 문제입니다.
> - 입력 데이터에 문제가 생길 때 발생하던 HttpMessageNotReadableException을 처리 못하고 있습니다.
> - BindingResult에 데이터가 없어도 에러가 발생할 수 있는 상황이 존재하는데 이런 상황 역시 BusinessException으로 핸들링 하도록 의도한 것인가요?

위와 같은 상황에서 상호간에 의견을 주고받으면 서로 응답을 기다리느라 그만큼 진행이 지연됩니다. 구현자의 이해도가 충분하지 않다는 가정 하에 리뷰와 함께 변경사항을 결정해서 통보하는 형태면 좋겠습니다. 대신에 리뷰 시 제안과 함께 이유를 조금 더 구체적으로 설명해 주면 좋겠습니다. 앞으로 제가 리뷰를 남길 때에도 이유를 상세히 적도록 주의하겠습니다.
> body가 json이 아닌 경우 HttpMessageNotReadableException이 발생하고,
이 경우 BindingResult가 null인데 이에 대한 처리가 누락되었다. BusinessException으로 처리해야 한다.

만약 어떤 의도를 가지고 구현한 내용이거나 다른 의견이 있다면 추가로 코멘트를 작성할 것이고, 그렇지 않다면 제안한 방법대로 구현할 것입니다.

앞으로도 다양한 의견을 제시할 상황이 생기겠지만, 최종 결정은 항상 main developer가 하는 형태였으면 합니다. harmony-server의 최종 담당자인 만큼, 이 저장소의 code에 한해서는 별도의 의논 없이 단독으로 처리해도 됩니다. approve 이후 직접 merge하지 않는 이유도 commit message나 branch 관리를 최종 결정권자가 할 수 있도록 하기 위함입니다. commit message 규칙이나 github flow, squash merge를 모두에게 소개했지만, 현재 각 저장소의 관리자에 따라 적용된 모양은 조금씩 다른 것을 확인할 수 있습니다. 앞으로도 메인 개발자의 동의 없이 변경사항이 적용되는 일은 없을 것이며, 메인 개발자의 결정권을 가능한 존중하고자 합니다. 이는 프로젝트의 진행 속도를 끌어올리기 위함이며, 개인적으로는 여러 사람의 프로젝트 관리 방법을 접하고 익숙해지기 위함입니다.

### 남은 변경사항
현재 시간 여유가 많지 않아 하모니 개발에 집중하기가 어렵습니다. 이 부분이 빠르게 처리되어야만 이후 작업을 진행할 수 있다면 우선 이대로 merge하고 남은 작업은 이슈로 등록해 이후에 처리하는 것이 좋겠습니다. 현재 코드도 동작에 큰 문제는 없는 상태이기 때문입니다. 꼭 필요한 변경사항이 있다면 본 PR을 직접 수정해도 됩니다.

namsic committed:
fix: Apply review

namsic commented:
@B2SIC
선택의 여지 없이 적용이 필요한 변경을 추가했습니다.

### 변경사항
- 추가 예외 처리
  - `HttpMessageNotReadableException`
  - `HttpMediaTypeNotSupportedException`
- 간단한 주석 추가
- Member
  - DTO 추가
  - password 조건 변경

B2SIC committed:
FIX: JPA Repository restore
FIX: Related code to repository
FEAT: Handling exceptions for empty parameter in find api
TEST: Modify test code associated with repository code

B2SIC reviewed (approved):
## 변경사항
### JPA 방식으로 복원
많은 고민 끝에 JPA 방식으로 복원했습니다. 하지만 궁극적으로는 Spring Data JPA 방식을 사용할 예정입니다. 
우선 현재 단계에서는 JPA로 진행하겠습니다. 인터페이스를 활용한 Data JPA 방식으로 데이터를 처리하면 물론 코드도 간결해져서 보기 좋고 쉬워진다는 이점도 있겠지만 그만큼 Spring Data JPA가 뒷단에서 처리해주는게 무엇인지 명확하게 보이지 않을 것 같다는 생각입니다. 따라서 JPA를 사용하면서 Spring Data JPA에 대한 이해를 거친 후 리팩토링을 통해 Spring Data JPA를 적용하도록 하겠습니다.

### Member Find API 예외 추가
Parameter로 userId와 phoneNumber 중에서 매칭되는 파라미터가 없을 경우 모든 데이터를 반환하는 방식에서 예외를 던지는 방식으로 변경했습니다.
현재 단계에서는 모든 데이터를 반환 할 일이 없을 것으로 보입니다.

### 첨언
Approve 상태로 두겠습니다.
리뷰 후 이상이 없으면 Merge 하고 작업을 이어가겠습니다.

B2SIC committed:
FIX: Null Exception Handling

B2SIC commented:
## 추가 변경사항
### Member Find API 버그 수정
파라미터로 아무것도 들어오지 않는 경우 이외에 매칭 되지 않는 경우에 대해서 처리가 안되던 문제 해결.

namsic commented on src/main/java/harmony/dev/harmonyserver/Exception/BusinessException.java:
```diff
@@ -0,0 +1,27 @@
+        return this.exceptions.add(e);
+    }
+
+    public void peekaboo() {
+        if(!this.exceptions.isEmpty()) throw this;
+    }
```
적당한 함수명이 떠오르지 않아 생각나는 이름을 사용했는데 의견이 있다면 변경 바랍니다.
Exception class에서 `throw this`와 같은 동작이 자연스러운지도 잘 모르겠지만,
마땅히 좋은 방법이 떠오르지 않아 이대로 두었습니다.

추가로 아래와 같은 메서드가 있으면
```java
public static void peekaboo(ExceptionSummary e) {
    this(e);
    throw this;
}
```
single error exception에서 아래와 같이 활용이 가능할 것 같은데 검토 바랍니다.
```java
if (userId == null && phoneNumber == null) {
    BusinessException.peekaboo(ExceptionSummary.builder()
                                               .code("NotFound")
                                               .build());
}
```
위 변경이 적용되면 `public BusinessException(ExceptionSummary e)`는 외부에서 호출할 필요가 없어 private으로 변경이 가능할 듯 합니다.

B2SIC replied:
이름은 그대로 사용하다가 적절한 이름이 생각나면 변경하겠습니다.
single error exception의 경우 언급 하신대로 반영할 경우 `BusinessException(ExceptionSummary e)` 의 기능을 `static void peekaboo(ExceptionSummary e)`가 대체 할 수 있지 않을까요?
적용해도 문제 없을 것으로 보입니다.

namsic commented on src/main/java/harmony/dev/harmonyserver/service/MemberService.java:
```diff
@@ -1,48 +1,67 @@
+            e.add(ExceptionSummary.builder()
+                            .field("userId or phoneNumber")
+                            .code("NotFound")
+                            .message("Parameter Not Found")
```
특정 field가 아닌 전반적인 문제라는 점에서 field를 null으로 두는 것은 어떤지 검토 바랍니다.
현재 구현도 크게 문제는 없을 것으로 보입니다.
```suggestion
                            .code("NotFound")
                            .message("No parameter given")
```

B2SIC replied:
반영해도 좋을 것 같습니다.
이름도 field 이기 때문에 message에 힌트를 주는 것으로 충분해보입니다.

namsic commented on src/main/java/harmony/dev/harmonyserver/SpringConfig.java:
```diff
@@ -1,34 +0,0 @@
-        return new MemberService(memberRepository());
-    }
-
-    @Bean
-    public MemberRepository memberRepository() {
-        return new DefaultMemberRepository(em);
-    }
```
해당 부분 사용하지 않아 SpringConfig.java 파일을 통째로 제거했는데, 복원하지 않아도 동작이 가능한가요?
두 방법 모두 동작 가능하다면 이후 관리 및 이해하기 편한 형태를 선택하는 것이 좋겠습니다.

B2SIC replied:
음.. 제가 알고 있는 대로라면 `SpringConfig.java`가 있어야 합니다..
지금 상태로는 의존성 주입이 안되는 것 맞을텐데 어떻게 DB에 데이터가 들어갔는지 모르겠네요.. 테스트에서도 문제 없어서 놓쳤던 것 같습니다.
원인을 좀 찾아봐야겠네요..

B2SIC replied:
컴포넌트 스캔에 의한 자동 의존관계 설정으로 `SpringConfig.java` 가 없어도 될 것 같습니다.

namsic commented on src/main/java/harmony/dev/harmonyserver/Exception/ExceptionAdvisor.java:
```diff
@@ -1,102 +1,56 @@
+
+    @ExceptionHandler(BusinessException.class)
+    @ResponseStatus(HttpStatus.CONFLICT)
+    protected ResponseDTO handleLogicalException(BusinessException ex) {
```
```suggestion
    protected ResponseDTO handleBusinessException(BusinessException ex) {
```

namsic commented on src/main/java/harmony/dev/harmonyserver/Exception/ExceptionAdvisor.java:
```diff
@@ -1,102 +1,56 @@
-    protected ErrorResponse notReadableException(HttpMessageNotReadableException ex){
-        return buildErrors(ErrorCode.BAD_REQUEST_TYPE);
+    protected ResponseDTO handleHttpMessageNotReadableException(HttpMessageNotReadableException ex) {
+        ArrayList<ExceptionSummary> errors = new ArrayList<>();
+        errors.add(ExceptionSummary.builder()
+                                   .code("UnreadableMessage")
+                                   .build());
+        return ResponseDTO.builder()
+                          .errors(errors)
+                          .build();
```
```suggestion
        return ResponseDTO.builder()
                          .errors(List.of(ExceptionSummary.builder()
                                                          .code("UnreadableMessage")
                                                          .build()))
                          .build();
```

namsic reviewed (commented):
LGTM

B2SIC committed:
Update src/main/java/harmony/dev/harmonyserver/service/MemberService.java

Co-authored-by: Namjae Kim <kimnj3050@gmail.com>

B2SIC committed:
Update src/main/java/harmony/dev/harmonyserver/Exception/ExceptionAdvisor.java

Co-authored-by: Namjae Kim <kimnj3050@gmail.com>

B2SIC committed:
Update src/main/java/harmony/dev/harmonyserver/Exception/ExceptionAdvisor.java

Co-authored-by: Namjae Kim <kimnj3050@gmail.com>

B2SIC commented on src/main/java/harmony/dev/harmonyserver/Exception/ExceptionSummary.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    public static List<ExceptionSummary> of(BusinessException businessException) {
+        return businessException.exceptions;
```
```suggestion
        return businessException.getExceptions();
```

namsic committed:
Update src/main/java/harmony/dev/harmonyserver/Exception/ExceptionSummary.java

Co-authored-by: Younghoon Kim <the_basic_@kookmin.ac.kr>

B2SIC committed:
FIX: Add List package
